--- beep.desktop.in_	2003-11-01 21:38:02.234608184 +0100
+++ beep.desktop.in	2003-11-01 21:40:45.765747672 +0100
@@ -1,8 +1,8 @@
 [Desktop Entry]
-Name=::beep.
-Comment=::beep. media player
-Exec=beep
-Icon=beep.svg
+Name=Beep
+Comment=Beep Media Player
+Exec=beep-media-player
+Icon=beep_mini.xpm
 MimeType=audio/x-scpls;audio/x-mpegurl;audio/mpegurl;audio/mp3;audio/x-mp3;audio/mpeg;audio/x-mpeg;audio/x-wav;application/x-ogg
 Terminal=0
 Type=Application
