--- dvbscanner.c.org	2003-05-31 09:51:26.000000000 +0200
+++ dvbscanner.c	2003-06-08 19:08:30.000000000 +0200
@@ -2,8 +2,8 @@
 #include <vdr/interface.h>
 #include <vdr/i18n.h>
 #include "dvbscanner.h"
-#include "../../../libdtv/libdtv.h"
-#include "../../../libdtv/libsi/si_debug_services.h"
+#include "/usr/include/vdr/libdtv.h"
+#include "/usr/include/vdr/si_debug_services.h"
 
 //#define DEBUG
 #ifdef DEBUG
