--- old/doc/lighttpd.conf	2004-04-05 09:46:55.000000000 +0200
+++ new/doc/lighttpd.conf	2004-04-05 09:54:45.828024776 +0200
@@ -25,10 +25,10 @@
 
 ## a static document-root, for virtual-hosting take look at the 
 ## server.virtual-* options
-server.document-root             = "/www/pages/"
+server.document-root             = "/home/lighttpd"
 
 ## where to send error-messages to
-server.errorlog            = "/www/logs/lighttpd.error.log"
+server.errorlog            = "/home/lighttpd/logs/lighttpd.error.log"
 
 # files to check for if .../ is requested
 server.indexfiles          = ( "index.php", "index.html", 
@@ -52,7 +52,7 @@
 				".conf" => "text/plain" )
 
 #### accesslog module
-accesslog.filename          = "/www/logs/access.log"
+accesslog.filename          = "/home/lighttpd/logs/access.log"
 
 ## deny access the file-extensions
 #
