--- Makefile.orig	2003-07-12 02:00:21.000000000 +0200
+++ Makefile	2003-07-12 02:15:49.000000000 +0200
@@ -16,7 +16,7 @@
 CXXFLAGS = -pipe -O2 -march=i586 -mcpu=i686 -fmessage-length=0 -fPIC -DNO_DEBUG -Wall -W -O2 -march=i586 -mcpu=i686 -fmessage-length=0 -fPIC -DNO_DEBUG  -DQT_NO_DEBUG
 LEXFLAGS = 
 YACCFLAGS= -d
-INCPATH  = -I$(QTDIR)/mkspecs/default -I. -I/usr/include -I$(QTDIR)/include -Ioutput/moc/
+INCPATH  = -I$(QTDIR)/mkspecs/linux-g++ -I. -I/usr/include -I$(QTDIR)/include -Ioutput/moc/
 LINK     = g++
 LFLAGS   = 
 LIBS     = $(SUBLIBS) -L/usr/lib/ -L$(QTDIR)/lib/ -L/usr/X11R6/lib/ -lqt -lXext -lX11 -lm
@@ -234,7 +234,7 @@
 $(MOC): 
 	( cd $(QTDIR)/src/moc ; $(MAKE) )
 
-Makefile: tuxcards.pro  $(QTDIR)/mkspecs/default/qmake.conf 
+Makefile: tuxcards.pro  $(QTDIR)/mkspecs/linux-g++/qmake.conf 
 	$(QMAKE) -o Makefile tuxcards.pro
 qmake: 
 	@$(QMAKE) -o Makefile tuxcards.pro
@@ -756,24 +756,24 @@
 ####### Install
 
 install_target: 
-	@$(CHK_DIR_EXISTS) $(INSTALL_ROOT)/usr/local/bin/ || $(MKDIR) $(INSTALL_ROOT)/usr/local/bin/
-	$(DEL_FILE) "$(INSTALL_ROOT)/usr/local/bin/$(QMAKE_TARGET)"
-	-$(COPY) "$(QMAKE_TARGET)" "$(INSTALL_ROOT)/usr/local/bin/$(QMAKE_TARGET)"
-	-strip "$(INSTALL_ROOT)/usr/local/bin/$(QMAKE_TARGET)"
+	@$(CHK_DIR_EXISTS) $(INSTALL_ROOT)/usr/bin/ || $(MKDIR) $(INSTALL_ROOT)/usr/bin/
+	$(DEL_FILE) "$(INSTALL_ROOT)/usr/bin/$(QMAKE_TARGET)"
+	-$(COPY) "$(QMAKE_TARGET)" "$(INSTALL_ROOT)/usr/bin/$(QMAKE_TARGET)"
+	-strip "$(INSTALL_ROOT)/usr/bin/$(QMAKE_TARGET)"
 
 uninstall_target: 
-	-$(DEL_FILE) "$(INSTALL_ROOT)/usr/local/bin/$(QMAKE_TARGET)"
-	-$(DEL_DIR) "$(INSTALL_ROOT)/usr/local/bin/"
+	-$(DEL_FILE) "$(INSTALL_ROOT)/usr/bin/$(QMAKE_TARGET)"
+	-$(DEL_DIR) "$(INSTALL_ROOT)/usr/bin/"
 
 
 install_documentation: 
-	@$(CHK_DIR_EXISTS) $(INSTALL_ROOT)/usr/local/doc/tuxcards/ || $(MKDIR) $(INSTALL_ROOT)/usr/local/doc/tuxcards/
-		-$(COPY_DIR) "src/gui/cactusbar/flowers" "$(INSTALL_ROOT)/usr/local/doc/tuxcards/"
+	@$(CHK_DIR_EXISTS) $(INSTALL_ROOT)/usr/doc/tuxcards/ || $(MKDIR) $(INSTALL_ROOT)/usr/doc/tuxcards/
+		-$(COPY_DIR) "src/gui/cactusbar/flowers" "$(INSTALL_ROOT)/usr/doc/tuxcards/"
 
 
 uninstall_documentation: 
-	-$(DEL_FILE) -r "$(INSTALL_ROOT)/usr/local/doc/tuxcards/src/gui/cactusbar/flowers"
-	-$(DEL_DIR) "$(INSTALL_ROOT)/usr/local/doc/tuxcards/"
+	-$(DEL_FILE) -r "$(INSTALL_ROOT)/usr/doc/tuxcards/src/gui/cactusbar/flowers"
+	-$(DEL_DIR) "$(INSTALL_ROOT)/usr/doc/tuxcards/"
 
 
 install: all install_target install_documentation 
