diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/Makefile vdr-1.2.6/work/vdr-1.2.6/Makefile
--- Original-vdr-1.2.6/work/vdr-1.2.6/Makefile	2003-12-13 16:05:18.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/Makefile	2003-12-13 16:08:31.000000000 +0100
@@ -14,23 +14,18 @@
 CXX      ?= g++
 CXXFLAGS ?= -g -O2 -Wall -Woverloaded-virtual
 
-DVBDIR   = ../DVB
 DTVDIR   = ./libdtv
-MANDIR   = /usr/local/man
-BINDIR   = /usr/local/bin
+MANDIR   = /usr/man
+BINDIR   = /usr/bin
 
-PLUGINDIR= ./PLUGINS
-PLUGINLIBDIR= $(PLUGINDIR)/lib
+PLUGINDIR= /usr/lib/vdr
+PLUGINLIBDIR= $(PLUGINDIR)
 
 VIDEODIR = /video
 
 DOXYGEN  = /usr/bin/doxygen
 DOXYFILE = Doxyfile
 
--include Make.config
-
-INCLUDES = -I$(DVBDIR)/include
-
 DTVLIB   = $(DTVDIR)/libdtv.a
 
 OBJS = audio.o channels.o ci.o config.o cutter.o device.o diseqc.o dvbdevice.o dvbosd.o\
@@ -135,13 +130,15 @@
 # Install the files:
 
 install:
-	@cp vdr runvdr $(BINDIR)
-	@gzip -c vdr.1 > $(MANDIR)/man1/vdr.1.gz
-	@gzip -c vdr.5 > $(MANDIR)/man5/vdr.5.gz
-	@if [ ! -d $(VIDEODIR) ]; then\
-            mkdir $(VIDEODIR);\
-            cp *.conf $(VIDEODIR);\
-            fi
+	@mkdir -p $(DESTDIR)$(BINDIR)
+	@mkdir -p $(BINDIR)
+	@cp vdr runvdr $(DESTDIR)$(BINDIR)
+	@mkdir -p $(DESTDIR)$(MANDIR)/man1
+	@mkdir -p $(DESTDIR)$(MANDIR)/man5
+	@mkdir -p $(MANDIR)/man1
+	@mkdir -p $(MANDIR)/man5
+	@gzip -c vdr.1 > $(DESTDIR)$(MANDIR)/man1/vdr.1.gz
+	@gzip -c vdr.5 > $(DESTDIR)$(MANDIR)/man5/vdr.5.gz
 
 # Source documentation:
 
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/Makefile vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/Makefile
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/Makefile	2003-12-13 16:05:18.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/Makefile	2003-12-13 16:08:31.000000000 +0100
@@ -40,7 +40,7 @@
 
 ### Includes and Defines (add further entries here):
 
-INCLUDES += -I$(VDRDIR)/include -I$(DVBDIR)/include
+INCLUDES += -I$(VDRDIR) -I$(DVBDIR)/include
 
 DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"'
 DEFINES += -D_GNU_SOURCE
@@ -69,7 +69,6 @@
 
 libvdr-$(PLUGIN).so: $(OBJS)
 	$(CXX) $(CXXFLAGS) -shared $(OBJS) -o $@
-	@cp $@ $(LIBDIR)/$@.$(VDRVERSION)
 
 dist: clean
 	@-rm -rf $(TMPDIR)/$(ARCHIVE)
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/hello.c vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/hello.c
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/hello.c	2002-12-13 16:01:43.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/hello.c	2003-12-13 16:08:31.000000000 +0100
@@ -8,8 +8,8 @@
 
 #include <getopt.h>
 #include <stdlib.h>
-#include <vdr/interface.h>
-#include <vdr/plugin.h>
+#include <interface.h>
+#include <plugin.h>
 #include "i18n.h"
 
 static const char *VERSION        = "0.1.0";
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/i18n.h vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/i18n.h
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/i18n.h	2002-05-11 16:48:16.000000000 +0200
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/hello/i18n.h	2003-12-13 16:08:31.000000000 +0100
@@ -9,7 +9,7 @@
 #ifndef _I18N__H
 #define _I18N__H
 
-#include <vdr/i18n.h>
+#include <i18n.h>
 
 extern const tI18nPhrase Phrases[];
 
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/osddemo/Makefile vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/osddemo/Makefile
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/osddemo/Makefile	2003-12-13 16:05:18.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/osddemo/Makefile	2003-12-13 16:08:31.000000000 +0100
@@ -40,7 +40,7 @@
 
 ### Includes and Defines (add further entries here):
 
-INCLUDES += -I$(VDRDIR)/include -I$(DVBDIR)/include
+INCLUDES += -I$(VDRDIR) -I$(DVBDIR)/include
 
 DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"'
 DEFINES += -D_GNU_SOURCE
@@ -69,7 +69,6 @@
 
 libvdr-$(PLUGIN).so: $(OBJS)
 	$(CXX) $(CXXFLAGS) -shared $(OBJS) -o $@
-	@cp $@ $(LIBDIR)/$@.$(VDRVERSION)
 
 dist: clean
 	@-rm -rf $(TMPDIR)/$(ARCHIVE)
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/osddemo/osddemo.c vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/osddemo/osddemo.c
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/osddemo/osddemo.c	2002-12-13 16:02:01.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/osddemo/osddemo.c	2003-12-13 16:08:31.000000000 +0100
@@ -6,7 +6,7 @@
  * $Id: vdr-1.2.6-gentoo.diff,v 1.2 2003/12/13 15:39:13 fow0ryl Exp $
  */
 
-#include <vdr/plugin.h>
+#include <plugin.h>
 
 static const char *VERSION        = "0.1.0";
 static const char *DESCRIPTION    = "Demo of arbitrary OSD setup";
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/sky/Makefile vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/sky/Makefile
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/sky/Makefile	2003-12-13 16:05:18.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/sky/Makefile	2003-12-13 16:08:31.000000000 +0100
@@ -40,7 +40,7 @@
 
 ### Includes and Defines (add further entries here):
 
-INCLUDES += -I$(VDRDIR)/include -I$(DVBDIR)/include
+INCLUDES += -I$(VDRDIR) -I$(DVBDIR)/include
 
 DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"'
 DEFINES += -D_GNU_SOURCE
@@ -69,7 +69,6 @@
 
 libvdr-$(PLUGIN).so: $(OBJS)
 	$(CXX) $(CXXFLAGS) -shared $(OBJS) -o $@
-	@cp $@ $(LIBDIR)/$@.$(VDRVERSION)
 
 dist: clean
 	@-rm -rf $(TMPDIR)/$(ARCHIVE)
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/sky/sky.c vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/sky/sky.c
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/sky/sky.c	2003-05-09 17:27:16.000000000 +0200
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/sky/sky.c	2003-12-13 16:08:31.000000000 +0100
@@ -11,8 +11,8 @@
 #include <sys/time.h>
 #include <sys/un.h>
 #include <unistd.h>
-#include <vdr/plugin.h>
-#include <vdr/sources.h>
+#include <plugin.h>
+#include <sources.h>
 
 static const char *VERSION        = "0.1.1";
 static const char *DESCRIPTION    = "Sky Digibox interface";
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/status/Makefile vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/status/Makefile
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/status/Makefile	2003-12-13 16:05:18.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/status/Makefile	2003-12-13 16:08:31.000000000 +0100
@@ -40,7 +40,7 @@
 
 ### Includes and Defines (add further entries here):
 
-INCLUDES += -I$(VDRDIR)/include -I$(DVBDIR)/include
+INCLUDES += -I$(VDRDIR) -I$(DVBDIR)/include
 
 DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"'
 DEFINES += -D_GNU_SOURCE
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/status/status.c vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/status/status.c
--- Original-vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/status/status.c	2002-12-13 16:01:53.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/PLUGINS/src/status/status.c	2003-12-13 16:08:31.000000000 +0100
@@ -6,8 +6,8 @@
  * $Id: vdr-1.2.6-gentoo.diff,v 1.2 2003/12/13 15:39:13 fow0ryl Exp $
  */
 
-#include <vdr/plugin.h>
-#include <vdr/status.h>
+#include <plugin.h>
+#include <status.h>
 
 static const char *VERSION        = "0.1.0";
 static const char *DESCRIPTION    = "Status monitor test";
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/plugin.c vdr-1.2.6/work/vdr-1.2.6/plugin.c
--- Original-vdr-1.2.6/work/vdr-1.2.6/plugin.c	2003-08-30 16:52:58.000000000 +0200
+++ vdr-1.2.6/work/vdr-1.2.6/plugin.c	2003-12-13 16:08:31.000000000 +0100
@@ -16,7 +16,7 @@
 #include "config.h"
 
 #define LIBVDR_PREFIX  "libvdr-"
-#define SO_INDICATOR   ".so."
+#define SO_INDICATOR   ".so"
 
 #define MAXPLUGINARGS  1024
 #define HOUSEKEEPINGDELTA 10 // seconds
@@ -283,7 +283,7 @@
   if (p)
      *p = 0;
   char *buffer = NULL;
-  asprintf(&buffer, "%s/%s%s%s%s", directory, LIBVDR_PREFIX, s, SO_INDICATOR, VDRVERSION);
+  asprintf(&buffer, "%s/%s%s%s", directory, LIBVDR_PREFIX, s, SO_INDICATOR);
   dlls.Add(new cDll(buffer, Args));
   free(buffer);
   free(s);
diff -Nru Original-vdr-1.2.6/work/vdr-1.2.6/vdr.c vdr-1.2.6/work/vdr-1.2.6/vdr.c
--- Original-vdr-1.2.6/work/vdr-1.2.6/vdr.c	2003-12-13 16:05:18.000000000 +0100
+++ vdr-1.2.6/work/vdr-1.2.6/vdr.c	2003-12-13 16:08:31.000000000 +0100
@@ -345,7 +345,8 @@
   // Configuration data:
 
   if (!ConfigDirectory)
-     ConfigDirectory = VideoDirectory;
+     ConfigDirectory = "/etc/vdr";
+//     ConfigDirectory = VideoDirectory;
 
   cPlugin::SetConfigDirectory(ConfigDirectory);
 
